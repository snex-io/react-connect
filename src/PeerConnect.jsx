import React, { Component, cloneElement } from "react";
import PropTypes from "prop-types";
import SNEX from "snex";

const MILLIS = 1000;
const GRACE = 30;

class PeerConnector extends Component {
  static propTypes = {
    onConnection: PropTypes.func.isRequired,
    type: PropTypes.string,
  };

  static defaultProps = {
    type: "nes",
  };

  constructor(props) {
    super(props);

    this.activating = false;
    this.session = null;
    this.timer = null;

    this.state = {
      busy: false,
      link: null,
    };
  }

  componentWillUnmount() {
    this.destroy();
  }

  componentWillReceiveProps;

  activate = () => {
    if (this.activating) {
      return;
    }

    this.activating = true;

    this.setState({
      busy: true,
      link: null,
    });

    if (!this.session) {
      this.session = SNEX.createSession();
      this.session
        .then(session => {
          session.on("connection", this.props.onConnection);
          session.on("disconnected", this.destroy);
        })
        .catch(e => {
          console.error(e);
          this.destroy();
        });
    }

    clearTimeout(this.timer);

    this.session
      .then(session => {
        return session.createURL(this.props.type);
      })
      .then(link => {
        this.activating = false;

        this.setState({
          link,
          busy: false,
        });

        const timeout = (link.lifeTime - GRACE) * MILLIS;
        console.info(`SNEX link timeout in ${timeout}ms`);

        this.timer = setTimeout(this.sleep, timeout);
      });
  };

  sleep = () => {
    this.setState({
      busy: false,
      link: null,
    });
  };

  destroy = () => {
    clearTimeout(this.timer);

    this.sleep();

    if (this.session) {
      this.session.then(session => {
        session.removeListener("connection", this.props.onConnection);
        session.removeListener("disconnected", this.destroy);
      });
      this.session = null;
    }
  };

  render() {
    const { busy, link } = this.state;
    const url = link ? link.url : null;

    return (
      <div className="snex-react-connect" onClick={this.activate}>
        {cloneElement(React.Children.only(this.props.children), { busy, url })}
      </div>
    );
  }
}

export default PeerConnector;
