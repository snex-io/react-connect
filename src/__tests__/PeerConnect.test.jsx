import React from "react";
import renderer from "react-test-renderer";

import { PeerConnect } from "../PeerConnect.jsx";

describe("PeerConnect", () => {
  let onConn;

  beforeEach(() => {
    onConn = jest.fn();
  });

  it('busy flag "false" and url "null" propagated by default', () => {
    const tree = renderer
      .create(
        <PeerConnect onConnection={onConn}>
          <div />
        </PeerConnect>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
