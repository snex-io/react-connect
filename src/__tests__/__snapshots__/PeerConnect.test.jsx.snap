// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeerConnect busy flag "false" and url "null" propagated by default 1`] = `
<div
  className="snex-react-connect"
  onClick={[Function]}
>
  <div
    busy={false}
    url={null}
  />
</div>
`;
